File: frontend/src/api/modules/point.ts
/**
 * 点位管理 API
 */
import request from '@/utils/request'
import type { PageParams, PageResponse } from './types'

export interface PointInfo {
  id: number
  point_code: string
  point_name: string
  point_type: 'AI' | 'DI' | 'AO' | 'DO'
  device_id: number
  device_type: string
  area_code: string
  unit: string
  data_type: string
  min_range: number
  max_range: number
  precision: number
  collect_interval: number
  store_interval: number
  is_enabled: boolean
  is_virtual: boolean
  calc_formula: string
  description: string
  sort_order: number
  created_at: string
  updated_at: string
}

export interface PointCreateParams {
  point_code: string
  point_name: string
  point_type: 'AI' | 'DI' | 'AO' | 'DO'
  device_id?: number
  device_type: string
  area_code: string
  unit?: string
  data_type?: string
  min_range?: number
  max_range?: number
  precision?: number
  collect_interval?: number
  store_interval?: number
  is_virtual?: boolean
  calc_formula?: string
  description?: string
  sort_order?: number
}

export interface PointUpdateParams extends Partial<PointCreateParams> {
  is_enabled?: boolean
}

export interface PointTypesSummary {
  total: number
  enabled: number
  AI: number
  DI: number
  AO: number
  DO: number
  by_area: Record<string, number>
  by_device_type: Record<string, number>
}

export interface PointGroup {
  id: number
  group_name: string
  group_type: 'area' | 'device_type' | 'custom'
  parent_id: number | null
  sort_order: number
  point_count: number
}

export interface PointGroupCreateParams {
  group_name: string
  group_type: 'area' | 'device_type' | 'custom'
  parent_id?: number
  sort_order?: number
  point_ids?: number[]
}

/**
 * 获取点位列表
 */
export function getPointList(params?: PageParams & {
  point_type?: string
  device_type?: string
  area_code?: string
  device_id?: number
  is_enabled?: boolean
  keyword?: string
}): Promise<PageResponse<PointInfo>> {
  return request.get('/v1/points', { params })
}

/**
 * 获取点位详情
 */
export function getPointById(id: number): Promise<PointInfo> {
  return request.get(`/v1/points/${id}`)
}

/**
 * 创建点位
 */
export function createPoint(data: PointCreateParams): Promise<PointInfo> {
  return request.post('/v1/points', data)
}

/**
 * 更新点位
 */
export function updatePoint(id: number, data: PointUpdateParams): Promise<PointInfo> {
  return request.put(`/v1/points/${id}`, data)
}

/**
 * 删除点位
 */
export function deletePoint(id: number): Promise<void> {
  return request.delete(`/v1/points/${id}`)
}

/**
 * 启用点位
 */
export function enablePoint(id: number): Promise<void> {
  return request.put(`/v1/points/${id}/enable`)
}

/**
 * 禁用点位
 */
export function disablePoint(id: number): Promise<void> {
  return request.put(`/v1/points/${id}/disable`)
}

/**
 * 批量导入点位
 */
export function batchImportPoints(data: FormData): Promise<{ success: number; failed: number; errors: string[] }> {
  return request.post('/v1/points/batch-import', data, {
    headers: { 'Content-Type': 'multipart/form-data' }
  })
}

/**
 * 导出点位配置
 */
export function exportPoints(params?: {
  point_type?: string
  device_type?: string
  area_code?: string
  format?: 'excel' | 'csv' | 'json'
}): Promise<Blob> {
  return request.get('/v1/points/export', {
    params,
    responseType: 'blob'
  })
}

/**
 * 获取点位类型统计
 */
export function getPointTypesSummary(): Promise<PointTypesSummary> {
  return request.get('/v1/points/types-summary')
}

/**
 * 获取点位分组
 */
export function getPointGroups(params?: { group_type?: string }): Promise<PointGroup[]> {
  return request.get('/v1/points/groups', { params })
}

/**
 * 创建点位分组
 */
export function createPointGroup(data: PointGroupCreateParams): Promise<PointGroup> {
  return request.post('/v1/points/groups', data)
}

File: backend/app/schemas/point.py
"""
点位相关 Schema
"""
from typing import Optional, Dict
from datetime import datetime
from pydantic import BaseModel, ConfigDict


class PointCreate(BaseModel):
    """创建点位"""
    point_code: str
    point_name: str
    point_type: str  # AI, DI, AO, DO
    device_id: Optional[int] = None
    device_type: str
    area_code: str
    unit: Optional[str] = None
    data_type: str = "float"
    min_range: Optional[float] = None
    max_range: Optional[float] = None
    precision: int = 2
    collect_interval: int = 10
    store_interval: int = 60
    is_virtual: bool = False
    calc_formula: Optional[str] = None
    description: Optional[str] = None


class PointUpdate(BaseModel):
    """更新点位"""
    point_name: Optional[str] = None
    device_id: Optional[int] = None
    device_type: Optional[str] = None
    area_code: Optional[str] = None
    unit: Optional[str] = None
    min_range: Optional[float] = None
    max_range: Optional[float] = None
    precision: Optional[int] = None
    collect_interval: Optional[int] = None
    store_interval: Optional[int] = None
    is_enabled: Optional[bool] = None
    description: Optional[str] = None
    sort_order: Optional[int] = None


class PointInfo(BaseModel):
    """点位信息"""
    model_config = ConfigDict(from_attributes=True)

    id: int
    point_code: str
    point_name: str
    point_type: str
    device_id: Optional[int] = None
    device_type: str
    area_code: str
    unit: Optional[str] = None
    data_type: str = "float"
    min_range: Optional[float] = None
    max_range: Optional[float] = None
    precision: int = 2
    collect_interval: int = 10
    store_interval: int = 60
    is_enabled: bool = True
    is_virtual: bool = False
    description: Optional[str] = None
    sort_order: int = 0
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None
    # 用能设备关联
    energy_device_id: Optional[int] = None
    energy_device_name: Optional[str] = None  # 关联的用能设备名称
    energy_device_code: Optional[str] = None  # 关联的用能设备编码


class PointTypesSummary(BaseModel):
    """点位类型统计"""
    total: int
    enabled: int
    ai: int = 0
    di: int = 0
    ao: int = 0
    do: int = 0
    by_device_type: Dict[str, int] = {}


class PointGroupCreate(BaseModel):
    """创建点位分组"""
    group_name: str
    group_type: Optional[str] = None
    parent_id: Optional[int] = None
    sort_order: int = 0


class PointGroupInfo(BaseModel):
    """点位分组信息"""
    model_config = ConfigDict(from_attributes=True)

    id: int
    group_name: str
    group_type: Optional[str] = None
    parent_id: Optional[int] = None
    sort_order: int = 0
    created_at: Optional[datetime] = None


# 保持向后兼容
PointBase = PointCreate
PointResponse = PointInfo


class PointRealtimeResponse(BaseModel):
    """点位实时数据响应"""
    model_config = ConfigDict(from_attributes=True)

    point_id: int
    point_code: str
    point_name: str
    point_type: str
    value: Optional[float] = None
    value_text: Optional[str] = None
    unit: Optional[str] = None
    quality: int = 0
    status: str = "normal"
    updated_at: Optional[datetime] = None


class PointHistoryResponse(BaseModel):
    """点位历史数据响应"""
    model_config = ConfigDict(from_attributes=True)

    id: int
    point_id: int
    value: float
    quality: int
    recorded_at: datetime


class PointTypeStats(BaseModel):
    """点位类型统计"""
    point_type: str
    count: int
    enabled_count: int


class RealtimeSummary(BaseModel):
    """实时数据汇总"""
    total: int
    normal: int
    alarm: int
    offline: int
